gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/pages/GradebookPage.java

```diff
@@ -155,6 +157,11 @@ public boolean isVisible() {
 		final List<Assignment> assignments = this.businessService.getGradebookAssignments();
 		Temp.time("getGradebookAssignments", stopwatch.getTime());
 
+		if (settings.isCategoriesEnabled()) {
+			// pre-sort assignments by the categorized sort order
+			Collections.sort(assignments, new CategorizedAssignmentComparator());
```

Comment:
Yeah, I think we could move this to the service to by extending the API of `getViewableAssignmentsForCurrentUser` to define a SortType and then by introducing a new SortType.SORT_BY_CATEGORY. We could then offering our fancy comparator in the service as a static on `Assignment` which would be used when SORT_BY_CATEGORY was chosen.  I've been hesitant to change too much in the service due to the fear of breaking Gradebook or Gradebook2!

Happy to have a go at the refactor if you think it's worth it?